File: julia/monte_carlo_single.jl
include("xoshiro256.jl")

const PI_THEORETICAL = 3.141592653589793238462643383279502884197

function calculate_pi(iterations::Int64)
    rng = Xoshiro256(UInt64(12345))
    inside_circle = 0
    
    for i in 1:iterations
        x = next_double(rng)
        y = next_double(rng)
        
        if x * x + y * y <= 1.0
            inside_circle += 1
        end
    end
    
    pi_estimate = 4.0 * inside_circle / iterations
    error = abs(pi_estimate - PI_THEORETICAL)
    
    return pi_estimate, error
end

function main()
    iterations = 100_000_000
    
    start = time()
    pi_estimate, error = calculate_pi(iterations)
    elapsed = time() - start
    
    elapsed_ms = elapsed * 1000.0
    
    println("{")
    println("  \"language\": \"Julia\",")
    println("  \"variant\": \"standard\",")
    println("  \"version\": \"$(VERSION)\",")
    println("  \"mode\": \"single\",")
    println("  \"iterations\": $iterations,")
    println("  \"pi_estimate\": $pi_estimate,")
    println("  \"error\": $error,")
    println("  \"time_ms\": $elapsed_ms,")
    println("  \"memory_mb\": 0.0,")
    println("  \"cache_misses\": 0,")
    println("  \"lines_of_code\": 0,")
    println("  \"compiler_flags\": \"N/A (JIT compiled)\",")
    println("  \"cpu_model\": \"N/A\",")
    println("  \"cpu_cores\": $(Sys.CPU_THREADS),")
    println("  \"thread_count\": 1,")
    println("  \"os\": \"$(Sys.KERNEL)\",")
    println("  \"os_version\": \"N/A\",")
    println("  \"compiler\": \"julia\",")
    println("  \"simd_detected\": false,")
    println("  \"simd_instructions\": []")
    println("}")
end

main()

File: go/monte_carlo_single.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"time"
)

const PI_THEORETICAL = 3.141592653589793238462643383279502884197

func calculatePi(iterations uint64) (float64, float64) {
	rng := NewXoshiro256(12345)
	insideCircle := uint64(0)

	for i := uint64(0); i < iterations; i++ {
		x := rng.NextDouble()
		y := rng.NextDouble()

		if x*x+y*y <= 1.0 {
			insideCircle++
		}
	}

	piEstimate := 4.0 * float64(insideCircle) / float64(iterations)
	error := piEstimate - PI_THEORETICAL
	if error < 0 {
		error = -error
	}

	return piEstimate, error
}

func main() {
	iterations := uint64(100000000)

	start := time.Now()
	piEstimate, error := calculatePi(iterations)
	elapsed := time.Since(start)

	elapsedMs := float64(elapsed.Nanoseconds()) / 1e6

	result := map[string]interface{}{
		"language":      "Go",
		"variant":       "standard",
		"version":       "1.21",
		"mode":          "single",
		"iterations":    iterations,
		"pi_estimate":   piEstimate,
		"error":         error,
		"time_ms":       elapsedMs,
		"memory_mb":     0.0,
		"cache_misses": 0,
		"lines_of_code": 0,
		"compiler_flags": "go build -ldflags=\"-s -w\"",
		"cpu_model":     "N/A",
		"cpu_cores":      runtime.NumCPU(),
		"thread_count":  1,
		"os":            runtime.GOOS,
		"os_version":    "N/A",
		"compiler":      "go",
		"simd_detected": false,
		"simd_instructions": []string{},
	}

	jsonData, _ := json.MarshalIndent(result, "", "  ")
	fmt.Println(string(jsonData))
	os.Exit(0)
}

